app/helpers/common_parameters_helper.rb

```diff
@@ -70,9 +70,31 @@ def lookup_key_field(id, value, options)
 
   def authorized_resource_parameters(resource, type)
     parameters_by_type = resource.send(type)
-    parameter_ids_to_view = parameters_by_type.authorized(:view_params).map(&:id)
-    resource_parameters = parameters_by_type.select { |p| parameter_ids_to_view.include?(p.id) }
+    @allowed_ids_hash = find_allowed_param_ids_per_action(parameters_by_type)
+    resource_parameters = parameters_by_type.select { |p| @allowed_ids_hash[:ids_to_view].include?(p.id) }
     resource_parameters += parameters_by_type.select(&:new_record?)
     resource_parameters
   end
+
+  def can_edit_parameter?(parameter)
+    @allowed_ids_hash[:ids_to_edit].include?(parameter.id)
+  end
+
+  def can_destroy_parameter?(parameter)
+    @allowed_ids_hash[:ids_to_destroy].include?(parameter.id)
+  end
+
+  def can_create_parameter?(parameter)
+    @can_create_param ||= authorizer.can?(:create_params)
+  end
+
+  private
+
+  def find_allowed_param_ids_per_action(parameters_by_type)
```

Comment:
I'm probably missins something and I'm reading this on phone but IIRC, authorizer exists exactly for this. It has separate caches for separate permissions. When it's instantiated, it can get preloaded set of objects. So I don't understand why we use it for create but bot for edit/destroy permissions here. If the collection of objects is different, we can have 3 authorizer instances.

Reply:
@ares, IIRC where this issue started from, the problem was that the cache is not filtered at all - you were getting all parameters from the DB, so it was too expensive to build that cache.

Reply:
@ShimShtein,

There are couple of changes suggested by @ares.
Basically, it includes instead of all parameters from the DB, it will limit to host's parameters.